festim/exports/txt_export.py

```diff
@@ -51,19 +57,58 @@ def is_it_time_to_export(self, current_time):
                 return True
         return False
 
-    def when_is_next_time(self, current_time):
-        if self.times is None:
-            return None
-        for time in self.times:
-            if current_time < time:
-                return time
-        return None
+    def is_last(self, current_time, final_time):
+        if final_time is None:
+            # write if steady
+            return True
+        elif self.times is None:
+            if np.isclose(current_time, final_time, atol=0):
+                # write at final time if exports at each timestep
+                return True
+        else:
+            if np.isclose(current_time, self.times[-1], atol=0):
+                # write at final time if exports at specific times
+                return True
+        return False
 
-    def write(self, current_time, steady):
-        # create a DG1 functionspace
-        V_DG1 = f.FunctionSpace(self.function.function_space().mesh(), "DG", 1)
+    def filter_duplicates(self, data, materials):
+        x = data[:, 0]
 
-        solution = f.project(self.function, V_DG1)
+        # Collect all borders
+        borders = []
+        for material in materials:
+            for border in material.borders:
+                borders.append(border)
+        borders = np.unique(borders)
+
+        # Find indices of the closest duplicates to interfaces
+        border_indx = []
+        for border in borders:
+            closest_indx = np.abs(x - border).argmin()
+            closest_x = x[closest_indx]
+            for ind in np.where(x == closest_x)[0]:
+                border_indx.append(ind)
+
+        # Find indices of first elements in duplicated pairs and mesh borders
+        _, unique_indx = np.unique(x, return_index=True)
+
+        # Combine both arrays of indices
+        combined_indx = np.concatenate([border_indx, unique_indx])
+
+        # Sort unique indices to return a slice
+        combined_indx = sorted(np.unique(combined_indx))
+
+        return data[combined_indx, :]
+
+    def write(self, current_time, final_time, materials, chemical_pot):
```

Comment:
This function now does a lot more than just "write" (it filters the data and has some logic etc).

Plus it seems like a lot of what is done here could be done only once.
For instance, `chemical_pot` will always be the same throughout the simulation, the materials will never change either!

Maybe we could put all this in a initialise step instead of doing this at each write?

Reply:
> Maybe we could put all this in a initialise step instead of doing this at each write?

somewhere [here](https://github.com/festim-dev/FESTIM/blob/a1e3aec1f73f39b339dedcf6946e2b5be13c8f6f/festim/generic_simulation.py#L390-L399)?

Reply:
Yes